### DISCUSSION TITLE
How to speedup clangd on large project? (around 2k idx files in cache)

### DISCUSSION THREAD

**Discussion by houtaru** [clangd/clangd#1601] [Q&A] [answered]
I have figured out that clangd generates around 2k files `*.idx` in the cache. it took a lot of time to load, and running updates. how to I speed it up?
Here are my clangd arguments:
```
 "clangd.arguments": [
        "--log=verbose",
        "--background-index",
        "--pretty",
        "-j=8",
        "--malloc-trim",
        "--pch-storage=memory",
        "--header-insertion=never",
	"--all-scopes-completion",
	"--clang-tidy",
	"--clang-tidy-checks=modernize-*,misc-*"
    ],
```

**Comment by HighCommander4** [comment]
Are you referring to the time it takes for clangd to build the index in the background? Or the time it takes for clangd to load the previously-built index into memory on startup? How long did it take, and how large is your project (# of source files)?

**Reply by houtaru** [reply]
I mean the time it takes for clangd to build the index in the background. It arounds 2-3s.
Here is an example: https://github.com/yugabyte/yugabyte-db

**Reply by HighCommander4** [reply]
Ok, it sounds like what you're actually talking about is the time is takes clangd to build the AST for a file after opening it in the editor, before features like semantic highlighting, go-to-definition etc. become available.

(Building the index refers to something different: clangd parsing all of the project's files in the background, and writing information to disk that will be used to support cross translation unit features like find-references. This can take minutes to hours depending on the size of the project, but since the results are persisted to disk it only needs to be done once.)

2-3s is actually quite good for AST build times. On a project I work on, it's not uncommon to see 5-10s.

**Reply by houtaru** [reply]
Would you like to share some tips to adjust clangd when you work on those projects? How are the arguments that I used?

**Reply by HighCommander4** [reply]
I haven't really experimented with `--pch-storage=memory` or `--malloc-trim`, so I can't attest to their effects on performance. I don't think the other options should have a performance impact, though it may not hurt to disable `--log=verbose` and `--pretty` unless actively investigating an issue that requires looking at logs.

**Reply by houtaru** [reply]
| On a project I work on, it's not uncommon to see 5-10s.
How do you configure clangd on these project?

**Reply by HighCommander4** [reply]
I don't do anything special related to performance. The only command-line flag I pass to clangd is `--compile-commands-dir` so it finds our `compile_commands.json` which is at a custom location (and also so that any headers located outside the project directory still interpolate command line flags from files in the project).

**Reply by houtaru** [reply]
Then, you wait for 5-10s each time you jump to another file in the project for clangd to run the update (then display on coc.clangd on vim)?

**Reply by HighCommander4** [reply]
Only the first time a given file is opened in the editing session. The expensive part of the AST build is building the "preamble" (basically, parsing the contents of the #included files), and preambles are retained over the lifetime of the editing session. So, yes, I face some 5-10s waits initially, but soon all of the files I commonly edit have had their preambles built and cached and switching back to any of those files is fast. (And I tend to keep my editor open to preserve this performant state where possible.)

**Reply by HighCommander4** [reply, marked as answer]
> preambles are retained over the lifetime of the editing session

Slight correction: the preamble for a file is retained as long as the file remains open in LSP (the client has sent a `textDocument/didOpen` and has not sent a `textDocument/didClose` for it). In vscode, this corresponds to the file being open in an editor. So, I do keep files open in background tabs to benefit from this performance gain, since if I close a file, its preamble is discarded.

I do think it would be a potentially nice improvement to retain preambles even for closed files; there is some discussion of this in https://github.com/clangd/clangd/issues/312 and https://github.com/clangd/clangd/discussions/1071.

I'm not sure how vim works, e.g. whether it has a concept of tabs. If it doesn't, and it actually closes the previous file with `textDocument/didClose` every time you switch files, that would be unfortunate for performance reasons and should probably be addressed in vim's LSP client.
